coalesce CIDRs across rules for l3 maps

ianvernon opened the pull request:
Add functionality in `pkg/ip` for combining sets of CIDRs into the smallest possible equivalent set of CIDR. Utilize this functionality for ingress / egress CIDR rules in Cilium.

Signed-off by: Ian Vernon <ian@cilium.io>

nebril commented on pkg/ip/ip.go:
```diff
@@ -26,6 +27,15 @@ const (
 	ipv6BitLen = 8 * net.IPv6len
 )
 
+var (
```
I think most of these vars could be represented as `net.IP` and `net.Mask` types.

ianvernon replied:
I prefer byte arrays because they are explicit and are not changed in the background - Go's `net` library manipulates the byte arrays you provide to `net.IP` and `net.Mask`.  Using byte arrays is what is done in the `net` package as well. See L55: https://golang.org/src/net/ip.go .

nebril replied:
Fair enough.

nebril commented on pkg/ip/ip.go:
```diff
@@ -54,10 +63,45 @@ func (s NetsByMask) Len() int {
+	}
+
+	// Then compare by CIDR.
+	//return less(s[i], s[j])
```
Please remove commented code and its comment.

nebril commented on pkg/ip/ip.go:
```diff
@@ -220,3 +262,424 @@ func flipNthBit(ip *[]byte, bitNum uint) *[]byte {
+
+func getPreviousIP(ip net.IP) net.IP {
+	// Cannot go lower than zero!
+	if bytes.Compare(defaultIPv4, ip) == 0 || bytes.Compare(defaultIPv6, ip) == 0 {
```
defaultIPv4.Equal(ip)? of course if you decide to represent vars as net.IP instead of []byte

nebril commented on pkg/ip/ip.go:
```diff
@@ -220,3 +262,424 @@ func flipNthBit(ip *[]byte, bitNum uint) *[]byte {
+
+	previousIP := make(net.IP, len(ip))
+	copy(previousIP, ip)
+	i := len(ip) - 1
```
nitpick: `i := len(ip) -1` could be moved to line 314 as `for` init statement

nebril commented on pkg/ip/ip.go:
```diff
@@ -220,3 +262,424 @@ func flipNthBit(ip *[]byte, bitNum uint) *[]byte {
+	}
+	nextIP := make(net.IP, len(ip))
+	copy(nextIP, ip)
+	i := len(ip) - 1
```
nitpick: `i := len(ip) -1` could be moved to line 342 as `for` init statement

nebril commented on pkg/ip/ip.go:
```diff
@@ -220,3 +262,424 @@ func flipNthBit(ip *[]byte, bitNum uint) *[]byte {
+}
+
+func getNextIP(ip net.IP) net.IP {
+	if bytes.Compare(upperIPv4, ip) == 0 || bytes.Compare(upperIPv6, ip) == 0 {
```
net.IP.Equal?

nebril commented on pkg/ip/ip.go:
```diff
@@ -220,3 +262,424 @@ func flipNthBit(ip *[]byte, bitNum uint) *[]byte {
+	return nextIP
+}
+
+func setNthBit(ip *[]byte, bitNum, val uint) *[]byte {
```
Why accept and return slice pointers? Won't references be enough here?

nebril commented on pkg/ip/ip.go:
```diff
@@ -220,3 +262,424 @@ func flipNthBit(ip *[]byte, bitNum uint) *[]byte {
+	lowestBig := big.NewInt(0).SetBytes(lowest)
+	highestBig := big.NewInt(0).SetBytes(highest)
+
+	// Starting from largest mask / smallest range possible, apply a mask one bit
```
Alternative would be to use native net.Mask type and do sth like that:

``` 
mask := net.CIDRMask(spanningMaskSize, bitLen)
for ; spanningMaskSize > 0 && !lowest.Mask(mask).Equal(highest.Mask(mask)); spanningMaskSize-- {
	mask = net.CIDRMask(spanningMaskSize, bitLen)
}
``` 

Not sure which is better/more performant though.

ianvernon replied:
I can add this as a TODO - would that be OK with you?

nebril replied:
Sure.

nebril commented on pkg/ip/ip.go:
```diff
@@ -220,3 +262,424 @@ func flipNthBit(ip *[]byte, bitNum uint) *[]byte {
+	// the IP network in the list itself
+	sort.Sort(NetsByRange(ranges))
+
+	// Merge adjacent CIDRs if possible.
```
Would be nice to have it extracted as a function.

ianvernon replied:
Done.

nebril reviewed (changes requested):
Looks really useful! Several nitpicks and one suggestion inline.

nebril commented on pkg/ip/ip_test.go:
```diff
@@ -168,3 +181,352 @@ func (s *IPTestSuite) TestByteFunctions(c *C) {
+	mergedV4CIDRs, mergedV6CIDRs = CoalesceCIDRs(cidrs)
+	s.testIPNetsEqual(mergedV4CIDRs, expected, c)
+
+	//v6 merge tests
```
Please remove.

nebril commented on tests/16-cidr-ingress-policy.sh:
```diff
@@ -70,6 +70,9 @@ log "IPV4_PREFIX: ${IPV4_PREFIX}"
 
 function test_cidr_except {
+  # This function tests importing policies with CIDR prefix exceptions
+  #not the datapath itself.
```
nitpick: add space after `#`

Commit:
pkg/ip: add functionality to coalesce CIDR list

Add CoalesceCIDRs and associated helper functions, which combines a list of
CIDRs into the smallest equivalent set of CIDRs.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
pkg/policy: utilize CoalesceCIDRs in Cilium

Merge CIDRs across all ingress / egress rules into the smallest equivalent set
of CIDRs.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
tests: update runtime tests for CIDR coalescing

Import rules which result in coalescing of CIDRs in runtime tests.

Signed-off by: Ian Vernon <ian@cilium.io>

joestringer commented on pkg/ip/ip_test.go:
```diff
@@ -61,10 +62,22 @@ func (s *IPTestSuite) testIPNetsEqual(created, expected []*net.IPNet, c *C) {
 
+func (s *IPTestSuite) testIPsEqual(created, expected net.IP, c *C) {
+	for k := range created {
+		c.Assert(reflect.DeepEqual(created[k], expected[k]), Equals, true)
```
I wonder if we should standardize using `comparator.DeepEquals()` rather than `reflect.DeepEqual`, so if anything like this ever fails then we get nice diff'd output. Maybe this PR just predates our comparator pkg.

ianvernon replied:
I'm going to move these tests to use `c.Assert(x, Equals, y)` as these are all primitive types and don't need `DeepEquals`. Will have that done in next PR.

joestringer commented on pkg/policy/rule.go:
```diff
@@ -327,6 +329,37 @@ func computeResultantCIDRSet(cidrs []api.CIDRRule) []api.CIDR {
 	return allResultantAllowedCIDRs
 }
 
+func cidrsToIPNets(cidrs []api.CIDR) (v4Nets, v6Nets []*net.IPNet) {
```
Unused function?

ianvernon replied:
Unused, but I left it around as it could prove useful.

joestringer commented on tests/16-cidr-ingress-policy.sh:
```diff
@@ -79,21 +82,59 @@ function test_cidr_except {
+        "fromCIDRSet": [ {
+            "cidr": "10.0.0.0/9",
             "except": [
-                "${IPV4_PREFIX_EXCEPT}"
```
I think this removes the last usage of `${IPV4_PREFIX_EXCEPT}`.

joestringer replied:
^ If this is unused, we should either be using it, or just delete it.

ianvernon replied:
I did delete it - see lines 70 and 65 which were removed.

joestringer replied:
Ah, perfect. I forget that GitHub doesn't detect stuff like that. Thanks!

joestringer commented on pkg/ip/ip_test.go:
```diff
@@ -168,3 +181,427 @@ func (s *IPTestSuite) TestByteFunctions(c *C) {
+	left, exclude, right = partitionCIDR(*targetCIDR, *excludeCIDR)
+	// Exclude should just contain exclude CIDR
+	s.testIPNetsEqual([]*net.IPNet{excludeCIDR}, exclude, c)
+	// Nothing should be in right list.
```
"left list". I saw a few more left/right mixups elsewhere in the PR too.

joestringer commented on pkg/policy/rule.go:
```diff
@@ -327,6 +329,37 @@ func computeResultantCIDRSet(cidrs []api.CIDRRule) []api.CIDR {
+	return
+}
+
+func (r *rule) mergeL3Ingress() {
```
Empty unused function?

joestringer commented on pkg/policy/rule_test.go:
```diff
@@ -550,12 +550,12 @@ func (ds *PolicyTestSuite) TestL3Policy(c *C) {
+	expected.Egress.Map["10.112.0.0/12"] = net.IPNet{IP: []byte{0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xff, 0xff, 10, 112, 0, 0}, Mask: []byte{255, 240, 0, 0}}
 	expected.Egress.IPv4Changed = true
-	expected.Egress.IPv4Count = 5
+	expected.Egress.IPv4Count = 4
```
Why does the expected egress IPv4 count go to 4?

ianvernon replied:
This is because the CIDRs are coalesced (10.1.0.0/16 is contained within one of the other CIDRs).

joestringer replied:
OK.

aanm replied:
Can you add that in a comment?

Also, you can create a new IPv4 with `net.IPv4()`

joestringer commented on tests/16-cidr-ingress-policy.sh:
```diff
@@ -111,6 +152,13 @@ wait_for_cilium_ep_gen
+
+
+test_cidr_except
+exit 0
```
Should this `exit 0` somewhere in the middle of the test like this?

joestringer reviewed (changes requested):
Looks good overall, I spotted a few unused functions and misc fixups that should be addressed before merge.

rlenglet commented on pkg/policy/l3.go:
```diff
@@ -38,45 +39,70 @@ type L3PolicyMap struct {
+// to the map.
+func (m *L3PolicyMap) Insert(cidrs []string) int {
+
+	ipNets := []*net.IPNet{}
```
You can pre-allocate this slice:
```
ipNets := make([]*net.IPNet, 0, len(cidrs) + len(m.Map))
```

rlenglet commented on pkg/policy/l3.go:
```diff
@@ -38,45 +39,70 @@ type L3PolicyMap struct {
+				mask = ip.DefaultMask() // IP address class based mask (/8, /16, or /24)
+				if !ip.Equal(ip.Mask(mask)) {
+					// IPv4 with non-zeroes after the subnetwork, use full mask.
+					mask = net.CIDRMask(32, 32)
```
You can just create the 32-long and 128-long masks once and for all. You shouldn't have to create them for each iteration.

rlenglet commented on pkg/policy/l3.go:
```diff
@@ -38,45 +39,70 @@ type L3PolicyMap struct {
+		key := ipnet.IP.String() + "/" + strconv.Itoa(ones)
+		if _, found := m.Map[key]; !found {
+			m.Map[key] = *ipnet
+			m.IPv6Count++
```
I'm intrigued by the semantics of these `IPv4Count` and `IPv6Count` fields. The resulting values don't make sense to me, as they are never set to zero anywhere. Shouldn't they be set to zero before the loops over `coalescedV4` and `coalescedV6`?
I think you should just do:
```
m.IPv4Count = len(coalescedV4)
m.IPv6Count = len(coalescedV6)
```

rlenglet commented on pkg/policy/l3.go:
```diff
@@ -38,45 +39,70 @@ type L3PolicyMap struct {
+		if _, found := m.Map[key]; !found {
+			m.Map[key] = *ipnet
+			m.IPv6Count++
+			m.IPv6Changed = true
```
`IPv4Changed` and `IPv6Changed` are never set to `false` anywhere. Shouldn't you do that in this method before the loops?
Those fields are important because they determine whether to regenerate BPF maps.
Computing those fields is non trivial. You should compare the resulting CIDR prefix map with the current `m.Map`.

ianvernon replied:
These fields were removed after this review.

rlenglet commented on pkg/policy/l3.go:
```diff
@@ -38,45 +39,70 @@ type L3PolicyMap struct {
+		ones, _ := ipnet.Mask.Size()
+
+		key := ipnet.IP.String() + "/" + strconv.Itoa(ones)
+		if _, found := m.Map[key]; !found {
```
Your code only adds prefixes into the map. This is not correct anymore because you're coalescing prefixes.
You should start from an empty map (`make(map[string]net.IPNet, len(coalescedV4) + len(coalescedV6))`, fill that map with the loops over `coalescedV4` and `coalescedV6`, and set `m.Map` to this new map at the end of the method.
This will have another advantage: keeping the old map intact will allow you to compare the old map with the new map, to determine whether anything has changed, and compute the `m.IPv4Changed` and `m.IPv6Changed` fields.

aanm commented on pkg/ip/ip.go:
```diff
@@ -26,6 +27,15 @@ const (
 	ipv6BitLen = 8 * net.IPv6len
 )
 
+var (
```
Can you add comments to them, I honestly don't know what the `v4Asv6` is supposed to be.

aanm commented on pkg/ip/ip.go:
```diff
@@ -38,8 +48,7 @@ func (s NetsByMask) Less(i, j int) bool {
 	if iPrefixSize == jPrefixSize {
-		byteArrComp := bytes.Compare(s[i].IP, s[j].IP)
-		if byteArrComp < 0 {
+		if bytes.Compare(s[i].IP, s[j].IP) < 0 {
```
You can even return the comparison :) `return bytes.Compare(s[i].IP, s[j].IP) < 0`

aanm commented on pkg/ip/ip.go:
```diff
@@ -220,3 +264,433 @@ func flipNthBit(ip *[]byte, bitNum uint) *[]byte {
+	lastIPMask := make(net.IPMask, len(ipNet.Mask))
+	copy(lastIPMask, ipNet.Mask)
+	for i := range lastIPMask {
+		lastIPMask[len(lastIPMask)-i-1] = ^lastIPMask[len(lastIPMask)-i-1]
```
What does this loop do?

aanm commented on pkg/ip/ip.go:
```diff
@@ -220,3 +264,433 @@ func flipNthBit(ip *[]byte, bitNum uint) *[]byte {
+
+func getPreviousIP(ip net.IP) net.IP {
+	// Cannot go lower than zero!
+	if ip.Equal(net.IP(defaultIPv4)) || ip.Equal(net.IP(defaultIPv6)) {
```
Isn't this line the same as `ip.IsUnspecified()`?

aanm commented on pkg/policy/rule.go:
```diff
@@ -336,23 +366,22 @@ func (r *rule) resolveL3Policy(ctx *SearchContext, state *traceState, result *L3
-		allCIDRs = append(allCIDRs, computeResultantCIDRSet(r.FromCIDRSet)...)
-
-		found += mergeL3(ctx, "Ingress", allCIDRs, &result.Ingress)
+	// L3 Ingress
```
Is this comment supposed to be `L3 Ingress` or it should be `L3 Egress`?

aalemayhu reviewed (commented):
Please ping when this is ready for review.

Commit:
WIP: review comments'

Commit:
WIP: fixing up unit tests
